### DISCUSSION TITLE
Override default inline css when compiling to wasm

### DISCUSSION THREAD

**Discussion by phrozen** [hajimehoshi/ebiten#1533] [Q&A]
The default wasm compliation creates the html, body and canvas with inline css style properties to a fixed 100% height. Is there a way to change those to a 100% width to allow for dynamically resizing the canvas? I'm currently using p5js embedded sketches in my dev blog and I can make the iframes resposive by changing the height after  a resize event. But for ebiten that does not work because the height is set initially and the canvas will just center vertically instead of resizing to the upper left corner.

**Comment by hajimehoshi** [comment]
Hi,

> But for ebiten that does not work because the height is set initially and the canvas will just center vertically instead of resizing to the upper left corner.

So you want to keep the screen on the upper left corner, right? What about modifying your Layout function to always return the outside size?

**Reply by phrozen** [reply]
Thanks for the reply! I'm loving this project, specially because it can compile to Wasm and I can embed projects. 

What will that make to the canvas once compiled to Wasm?

I checked the code and I see the html and body are set to 100% height and that is taken from the initial iframe size. The problem is once that is done, I cannot resize the iframe with the canvas inside. 

What p5js does on the other hand (I actually can hack that part) is that is set to width 100% and the height is calculated based on that, in html, it is more common for elements to have a width than a height. If the canvas is set for width 100% I can then resize my content and easily fix the height with javascript to make the content responsive.

Let me code a quick example later today to show you what I mean.

**Reply by phrozen** [reply]
@hajimehoshi Here is the example repo, code and demo. Hopefully it is more clear what I'm trying to achieve.

Demo: https://phrozen.github.io/canvas-embed-example/
Repo: https://github.com/phrozen/canvas-embed-example

I will greatly appreciate your help to make sure I can hack the ebiten part to match full width of it's parent.

**Reply by hajimehoshi** [reply]
As I mentioned, please check your Layout. How to scale the game screen to the outside size (iframe size in this case) depends on Layout. You're now returning a game's logical screen size at Layout, but I'm suggesting to return the outside size if you want to avoid the automatic scaling by Ebiten. This would require some modify in your game logic.

See also https://ebiten.org/blog/resizable.html

**Reply by phrozen** [reply]
I understand how Layout works, and it works as expected when compiling for desktop. But for wasm, the width and height of the canvas itself is hard coded to the body at the moment of resizing, not to my layout values:

https://github.com/hajimehoshi/ebiten/blob/a0a8d41ff63f4d942bd01df078c95dcbed749248/internal/uidriver/js/ui_js.go#L471

It doesn't matter what I return in Layout, the canvas size will always be the body width and height. I want the canvas to be the size of my logic screen so the browser can scale the canvas for me. It makes sense to me that in web, Layout determines the size of the canvas. So that:

> If Layout returns ignores the outside size and returns a fixed size, the game screen size is not changed but the screen scale is adjusted instead.

Maybe there is something I'm missing, do you have a very basic example where a logical screen project of arbitrary `width, height` scales to a 100% of it's parent element using wasm? For me I can't make it work because the inline properties of the canvas do not depend on my code, but on the container (iframe) itself, so to set the width and height I have to know them always before hand. 

Thanks you so much for taking time for this.

One approach that I think will let me do it, is that `SetWindowSize` does nothing on wasm, but if that could be wired to set the  **fixed** `width, height` of the canvas attributes, it would be awesome! That way I can keep the canvas at the size of my project screen (in Go code), and just let the browser scale the canvas as I see fit with CSS. The thing with canvas is that the attributes `width, height` are the size of the screen in pixels, and then you can scale it to whatever you want using CSS without changing the size of the actual canvas and pixel count.

**Reply by hajimehoshi** [reply]
>  But for wasm, the width and height of the canvas itself is hard coded to the body at the moment of resizing, not to my layout values
> [...]
> It doesn't matter what I return in Layout, the canvas size will always be the body width and height.

Yes, and the scaling is done by Ebiten automatically. If `Layout` returns the outside size, the scaling is not done (to be exact, Ebiten considers `devicePixelRatio` values, so some slight scaling is done).

> Maybe there is something I'm missing, do you have a very basic example where a logical screen project of arbitrary width, height scales to a 100% of it's parent element using wasm? 

[`examples/windowsize`](https://github.com/hajimehoshi/ebiten/blob/master/examples/windowsize/main.go) is a good example. Unfortunately, this require `-autoadjusting` flag to see Layout returning the outside size, so it might be hard to see this on Wasm. You can check this on desktops easily.

Anyway, I recommend to rewrite Layout like this

```
func (game *GameOfLife) Layout(width, height int) (int, int) {
	return width, height
}
```

and see what's going on.

> For me I can't make it work because the inline properties of the canvas do not depend on my code, but on the container (iframe) itself, so to set the width and height I have to know them always before hand.

Sorry but I don't understand why you need to know the values before hand.

Probably I don't fully understand what you want to do... Could you show figures explaining how you want to make your game work?

**Reply by hajimehoshi** [reply]
I mean I want to know how you want to make your app responsive. I still believe fixing Layout would solve this.

**Reply by hajimehoshi** [reply]
> (returning the outside size in Layout will just move the tiny canvas to the upper left corner)

Probably you want to use a fixed logical width value and a flexible height, right? What about:

```
func (game *GameOfLife) Layout(width, height int) (int, int) {
	return fixedWidth, int(fixedWidth * float64(height) / float64(width))
}
```

**Reply by phrozen** [reply]
I'm trying every solution you suggest without success, so I think is better I explain the use case and then there might be a better way to handling it.

I have a blog (done with Hugo), I put a sketches/projects where I do generative art and stuff like that. This blog's main content has a `max-width` lets say of `X` and it is responsive, so in mobile and tablets or whatever, the content scales based on the width.

What I want is to be able to embed `ebiten` demos and projects in wasm with an `<iframe>` so that no matter what the size of the logical screen is inside the demo, it shows always at full width and adjusted height (maintaining aspect ratio), even after the browser is resized.

The `iframe` width is set to `100%` so that it always fits the container, and right now for p5.js demos I'm setting the height with Javascript after the canvas is rendered with the iframe `onload` event.

So, if for example I create a project where the logical screen is `180x120` as in the example demo, I want it to be scaled up at the full width of the `iframe`, pixelated of course, and adjust the height accordingly (automatically or I can do it via JS). 

And if I have another demo with a logical screen of `1500x3000` I want it to be scaled down to fit the width of the iframe and adjust the height accordingly, even after resizing the browser window below the `max-width`. But because I'm using shortcodes to generate the iframes and such, I don't want to set the `iframe` width and height explicitly.

Here is an example with the p5.js sketches on the blog (near the bottom): https://gestrada.dev/posts/the-game-of-life/

Try resizing the browser window and see how no matter what, the `iframe` keeps the aspect ratio and resizes accordingly. Also using the fullscreen button works as expected.

I hope it is more clear what I need, I don't mind doing some javascript to make it work, it's just I can't figure it out with just ebiten.

Thank you!

**Reply by hajimehoshi** [reply]
Thank you for clarifying! So my question

> Probably you want to use a fixed logical width value and a flexible height, right?

sounds right, but didn't my last suggestion work?

What logical screen size do you want? Is the below table correct?

outside size|logical screen size
---|---
180x120|180x120
1500x3000|180x360

In theory, any complex layout logic can be done in Layout.

**Reply by phrozen** [reply]
I kind of figured it out. It has nothing to do with the layout. That will be as is (logical screen size) fixed values.

What I figured out by playing with the dev tools is that Ebiten changes the internal (embedded) canvas size based on the iframe width and height automatically!

I really wish there was a way for the ebiten canvas width and height to remain constant though. But if it is impossible, I just need a way to change the iframe width and height dynamically based on it's parent. That might be doable, although a bit of a hack.

I want to keep my logical screen size always the same (same pixel count). And scale it to the full width of the container (iframe) and maintain aspect ratio on the height. Dinamically/responsive.

Hope that makes sense.

**Reply by phrozen** [reply]
In the end, this is kinda working:

```js
function resizeEbiten(iframe) {
    let w = document.getElementById('iframe').clientWidth; // this is the parent div containing the iframe
    iframe.width = w;
    iframe.height = w * 120 / 180; //will have to find a way to maintain aspect ratio
    iframe.contentWindow.addEventListener('resize', () => {
        let w = document.getElementById('iframe').clientWidth;
        iframe.width = w;
        iframe.height = w * 120 / 180;
    });
}
```

But because Ebiten is really hardcoded/wired to the iframe's width and height, I still need to know the aspect ratio of the demo beforehand...

**Reply by hajimehoshi** [reply]
> What I figured out by playing with the dev tools is that Ebiten changes the internal (embedded) canvas size based on the iframe width and height automatically!

Yes, updateSize is called anytime when the iframe size is changed.

> I want to keep my logical screen size always the same (same pixel count). And scale it to the full width of the container (iframe) and maintain aspect ratio on the height. Dinamically/responsive.

So did my last suggestion work?

```
func (game *GameOfLife) Layout(width, height int) (int, int) {
	return fixedWidth, int(fixedWidth * float64(height) / float64(width))
}
```

**Reply by phrozen** [reply]
Unfortunately no, there is no way around it. I want to keep my logical screen size always the same.

> Yes, updateSize is called anytime when the iframe size is changed.

The problem is that to make it responsive, Ebiten only changes the canvas size based on the iframe width and size. So the only way to actually resize Ebiten's canvas, is to resize the iframe itself.

Here is the demo with the working solution by recalculating the iframe's width and height (not ideal but works), try resizing the browser to see the difference between examples:

https://phrozen.github.io/canvas-embed-example/#solution

**Reply by phrozen** [reply]
It is also REALLY weird and this is probably browser related...

If I have fixed width and height iframe, and I resize the window in Chrome (windows), only the width changes and the height remains the same.

![image](https://user-images.githubusercontent.com/213238/110910121-89b6c180-82d6-11eb-925a-5d46ebbd5a59.png)

But if I use the Dev Tools "Responsive" mode with "toggle Device Toolbar" then the height changes also when resizing! (which is what I wanted in the first place!

![image](https://user-images.githubusercontent.com/213238/110910366-d4d0d480-82d6-11eb-9c88-abf92afa69d5.png)

Is that a bug?

**Reply by hajimehoshi** [reply]
Aha, perhaps do you want to keep the aspect ratio of the iframe? In this case, this is not Ebiten's responsibility since Ebiten doesn't care what the outside world is.

IIUC, in CSS, just setting max-width and not setting width/height should keep the aspect ratio of the iframe element.

https://stackoverflow.com/questions/12912048/how-to-maintain-aspect-ratio-using-html-img-tag

**Reply by phrozen** [reply]
Thank you! Will take a look!
